ethers-middleware/src/multicall.rs

```diff
@@ -0,0 +1,386 @@
+#![cfg(not(target_arch = "wasm32"))]
+
+use std::{ops::Deref, sync::Arc};
+
+use async_trait::async_trait;
+use ethers_contract::{
+    multicall::{
+        contract::{
+            GetBlockNumberCall, GetBlockNumberReturn, GetEthBalanceCall, GetEthBalanceReturn,
+            MULTICALL3_ABI,
+        },
+        Multicall,
+    },
+    BaseContract, ContractCall, ContractError, MulticallError,
+};
+use ethers_core::{
+    abi::{encode, Abi, AbiDecode, AbiEncode, Token, Tokenizable},
+    types::{
+        transaction::eip2718::TypedTransaction, Address, BlockId, Bytes, NameOrAddress,
+        TransactionRequest,
+    },
+};
+use ethers_providers::{Middleware, MiddlewareError};
+use thiserror::Error;
+
+use tokio::sync::{mpsc, oneshot};
+
+type MulticallResult<M> = Result<Token, Arc<MulticallError<M>>>;
+type MulticallRequest<M> = (ContractCall<M, Token>, oneshot::Sender<MulticallResult<M>>);
+
+#[derive(Debug)]
+/// Processor for multicall middleware requests
+pub struct MulticallProcessor<M: Middleware> {
+    inner: Arc<M>,
+    multicall_address: Option<Address>,
+    max_batch_size: usize,
+    rx: mpsc::Receiver<MulticallRequest<M>>,
+}
+
+#[derive(Debug, Clone)]
+/// Middleware used for transparently leveraging multicall functionality
+pub struct MulticallMiddleware<M: Middleware> {
+    inner: Arc<M>,
+    contracts: Vec<BaseContract>,
```

Comment:
this is immutable?
can Arc then?

Reply:
I was thinking that maybe when new contracts are instantiated we might append here?